src/Google.Api.Gax/Api/Gax/PathTemplate.cs

```diff
@@ -0,0 +1,439 @@
+﻿/*
+ * Copyright 2016 Google Inc. All Rights Reserved.
+ * Use of this source code is governed by a BSD-style
+ * license that can be found in the LICENSE file or at
+ * https://developers.google.com/open-source/licenses/bsd
+ */
+
+using System;
+using System.Collections.Generic;
+using System.Collections.ObjectModel;
+using System.Linq;
+using System.Text;
+
+namespace Google.Api.Gax
+{
+    /// <summary>
+    /// Represents a path template used for resource names which may be composed of multiple IDs.
+    /// </summary>
+    /// <remarks>
+    /// <para>
+    /// Templates use a subset of the syntax of the API platform. See the protobuf of <see cref="HttpRule"/> for
+    /// details of the API platform.
+    /// </para>
+    /// <para>
+    /// This class performs no URL escaping or unescaping. It is designed for use within GRPC, where no
+    /// URL encoding is required.
+    /// </para>
+    /// </remarks>
+    public sealed class PathTemplate
+    {
+        /// <summary>
+        /// Just an array containing a single slash, to avoid constructing a new array every time we need
+        /// to split.
+        /// </summary>
+        private static readonly char[] SlashSplit = { '/' };
+
+        /// <summary>
+        /// List of segments in this template. Never modified after construction.
+        /// </summary>
+        private readonly IList<Segment> _segments;
+        /// <summary>
+        /// List of the segments in this template which are wildcards. Never modified after construction.
+        /// </summary>
+        private readonly IList<Segment> _parameterSegments;
+
+        private readonly string _originalTemplate;
+        private readonly bool _hasPathWildcard;
+
+        /// <summary>
+        /// The names of the parameters within the template. This collection has one element per parameter,
+        /// but unnamed parameters have a name of <c>null</c>.
+        /// </summary>
+        public ReadOnlyCollection<string> ParameterNames { get; }
```

Comment:
After a bit of discussion, Ivan and I reckoned that `IReadOnlyList<T>` might be a better option after all, just so that we can update to immutable collections at a later date.